src/main/scala/viper/gobra/translator/encodings/structs/StructEncoding.scala

```diff
@@ -0,0 +1,243 @@
+    * R[ dflt(Struct{F}) ] -> create_ex_struct( [T] | (f: T) in F )
+    * R[ structLit(E) ] -> create_ex_struct( [e] | e in E )
+    * R[ loc: Struct{F}@ ] -> convert_to_exclusive( Ref[loc] )
+    */
+  override def expr(ctx: Context): in.Expr ==> CodeWriter[vpr.Exp] = default(super.expr(ctx)){
+    case (loc@ in.FieldRef(recv :: ctx.Struct(fs), field)) :: _ / Exclusive =>
+      for {
+        vBase <- ctx.expr.translate(recv)(ctx)
+        idx = fs.indexOf(field)
+      } yield ex.get(vBase, idx, cptParam(fs)(ctx))(loc)(ctx)
+
+    case (upd: in.StructUpd) :: ctx.Struct(fs) =>
+      for {
+        vBase <- ctx.expr.translate(upd.base)(ctx)
+        idx = fs.indexOf(upd.field)
+        vVal <- ctx.expr.translate(upd.newVal)(ctx)
+      } yield ex.update(vBase, idx, vVal, cptParam(fs)(ctx))(upd)(ctx)
+
+    case (e: in.DfltVal) :: ctx.Struct(fs) / Exclusive =>
+      val fieldDefaults = fs.map(f => in.DfltVal(f.typ)(e.info))
+      sequence(fieldDefaults.map(ctx.expr.translate(_)(ctx))).map(ex.create(_, cptParam(fs)(ctx))(e)(ctx))
+
+    case (e: in.DfltVal) :: ctx.Struct(fs) / Shared =>
+      val (pos, info, errT) = e.vprMeta
+      unit(shDfltFunc(Vector.empty, fs)(pos, info, errT)(ctx))
+
+    case (lit: in.StructLit) :: ctx.Struct(fs) =>
+      val fieldExprs = lit.args.map(arg => ctx.expr.translate(arg)(ctx))
+      sequence(fieldExprs).map(ex.create(_, cptParam(fs)(ctx))(lit)(ctx))
+
+    case (loc: in.Location) :: ctx.Struct(_) / Shared =>
+      sh.convertToExclusive(loc)(ctx, ex)
+  }
+
+  /**
+    * Encodes the reference of an expression.
+    *
+    * To avoid conflicts with other encodings, an encoding for type T should be defined at shared operations on type T.
+    * Super implements shared variables with [[variable]].
+    *
+    * Ref[ (e: Struct{F}@).f ] -> sh_struct_get(Ref[e], f, F)
+    */
+  override def reference(ctx: Context): in.Location ==> CodeWriter[vpr.Exp] = default(super.reference(ctx)){
+    case (loc@ in.FieldRef(recv :: ctx.Struct(fs), field)) :: _ / Shared =>
+      for {
+        vBase <- ctx.typeEncoding.reference(ctx)(recv.asInstanceOf[in.Location])
+        idx = fs.indexOf(field)
+      } yield sh.get(vBase, idx, cptParam(fs)(ctx))(loc)(ctx)
+  }
+
+  /**
+    * Encodes the permissions for all addresses of a shared type,
+    * i.e. all permissions involved in converting the shared location to an exclusive r-value.
+    * An encoding for type T should be defined at all shared locations of type T.
+    */
+  override def addressFootprint(ctx: Context): in.Location ==> CodeWriter[vpr.Exp] = {
+    case loc :: ctx.Struct(_) / Shared => sh.addressFootprint(loc)(ctx)
+  }
+
+  /** Returns 'base'.f for every f in 'fields'. */
+  private def fieldAccesses(base: in.Expr, fields: Vector[in.Field]): Vector[in.FieldRef] = {
+    fields.map(f => in.FieldRef(base, f)(base.info))
+  }
+
+  /**
+    * Generates:
+    * function shStructDefault(): [Struct{F}@]
+    *   ensures AND (f: T) in F. [&result.f == dflt(T)]
```

Comment:
why `AND`? Shouldn't it be `forall`?

Reply:
Since it is effectifely iterated, I used AND.